### DISCUSSION TITLE
Compute volume between two open meshes

### DISCUSSION THREAD

**Discussion by pzaffino** [pyvista/pyvista#3392] [Q&A]
Hello,
I have two open meshes (collected by a scanner) and I would like to compute the volume between them (in case adopting some strategy to "close the volume between them").

Any idea?

Thank you very much.

Best,
Paolo

**Comment by banesullivan** [comment]
This is going to be a bit difficult since the meshes are open. You'll have to enclose the volume between the two open surfaces somehow - this could be done a lot of different ways depending on your particular dataset.

I'd recommend using either some sort of extrusion filter or use something like [PyMeshFix](https://pymeshfix.pyvista.org/) to repair the surfaces or both.

See

- https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.extrude.html?highlight=extru#pyvista.PolyDataFilters.extrude
- https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.extrude_trim.html?highlight=extru#pyvista.PolyDataFilters.extrude_trim
- https://docs.pyvista.org/examples/01-filter/extrude-trim.html#sphx-glr-examples-01-filter-extrude-trim-py

For further volume analysis after you have a closed surface, take a look at:

- https://docs.pyvista.org/examples/01-filter/compute-volume.html#sphx-glr-examples-01-filter-compute-volume-py
- https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSet.volume.html?highlight=volume#pyvista.DataSet.volume
- https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSetFilters.compute_cell_sizes.html?highlight=cell+sizes#pyvista.DataSetFilters.compute_cell_sizes

**Comment by pzaffino** [comment]
Thanks a lot @banesullivan !
I have two additional questions about volume calculation:
1) is the "volume" attribute measured in cm^3 or in mm^3?
2) since the "volume" attribute has a value also in the case of open mesh, how it is computed?

Thanks a lot.
Paolo

**Reply by banesullivan** [reply]
>is the "volume" attribute measured in cm^3 or in mm^3?

PyVista is unitless - whatever units your XYZ coordiantes are in will be the units of this volume calculation. If your axes are in cm then the volume would be cm^3

> since the "volume" attribute has a value also in the case of open mesh, how it is computed?

We use `vtkMassProperties` to compute the volume: https://vtk.org/doc/nightly/html/classvtkMassProperties.html#details

> The algorithm implemented here is based on the discrete form of the divergence theorem. The general assumption here is that the model is of closed surface

Is the volume non-zero for an open PolyData mesh?

**Reply by pzaffino** [reply]
Got it!
Thank you

**Comment by pzaffino** [comment]
Sorry, maybe I was not clear.
If I provide an open surface and still I get an output value, how does the algorithm close the volume?

**Reply by banesullivan** [reply]
Generally, you should not trust the volume estimation for a non-closed surface. The algorithm assumes it is a close surface. (See the `vtkMassProperties` link I shared above).

We recommend checking if a mesh is closed before computing the volume using the `is_manifold` property on `PolyData` objects: https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.is_manifold.html?highlight=manifold#pyvista.PolyData.is_manifold
